cudax/include/cuda/experimental/__stf/internal/cuda_kernel_scope.cuh

```diff
@@ -45,46 +67,207 @@ class stream_ctx;
+  }
+
+  template <typename Fun, typename... Args>
+  void configure(Fun func, dim3 gridDim_, dim3 blockDim_, size_t sharedMem_, Args... args)
   {
     using TupleType = ::std::tuple<::std::decay_t<Args>...>;
 
+    _CCCL_ASSERT(!configured, "cuda_kernel_desc was already configured");
+
+    func_variant = store_func(mv(func));
+    gridDim      = gridDim_;
+    blockDim     = blockDim_;
+    sharedMem    = sharedMem_;
+
     // We first copy all arguments into a tuple because the kernel
     // implementation needs pointers to the argument, so we cannot use
     // directly those passed in the pack of arguments
-    auto arg_tuple = ::std::make_shared<TupleType>(std::forward<Args>(args)...);
+    auto arg_tuple = ::std::make_shared<TupleType>(mv(args)...);
 
-    // Ensure we are packing arguments of the proper types to call func
-    static_assert(::std::is_invocable_v<Fun, Args...>);
+    // Ensure we are packing arguments of the proper types to call func (only
+    // valid with the runtime API)
+    if constexpr (!reserved::is_function_or_kernel_v<Fun>)
+    {
+      static_assert(::std::is_invocable_v<Fun, Args...>);
+    }
 
     // Get the address of every tuple entry
     ::std::apply(
       [this](auto&... elems) {
         // Push back the addresses of each tuple element into the args vector
-        ((args_ptr.push_back(static_cast<void*>(&elems))), ...);
+        ((args_ptr.push_back(&elems)), ...);
       },
       *arg_tuple);
 
     // Save the tuple in a typed erased value
     arg_tuple_type_erased = mv(arg_tuple);
+
+    configured = true;
+  }
+
+  // It is the responsibility of the caller to unsure arguments are valid until
+  // the CUDA kernel construct ends
+  template <typename Fun>
+  void configure_raw(Fun func, dim3 gridDim_, dim3 blockDim_, size_t sharedMem_, int arg_cnt, const void** args)
+  {
+    _CCCL_ASSERT(!configured, "cuda_kernel_desc was already configured");
+
+    func_variant = store_func(mv(func));
+    gridDim      = gridDim_;
+    blockDim     = blockDim_;
+    sharedMem    = sharedMem_;
+
+    for (int i = 0; i < arg_cnt; i++)
+    {
+      // We can safely forget the const here because CUDA will not modify the
+      // argument
+      args_ptr.push_back(const_cast<void*>(args[i]));
+    }
+
+    configured = true;
   }
 
-  /* __global__ function */
-  const void* func;
+  /* CUfunction/CUkernel (CUDA driver API) or __global__ function (CUDA runtime API) */
+  using func_variant_t =
+    ::std::variant<CUfunction,
+#if CUDA_VERSION >= 12000
+                   CUkernel,
+#endif
+                   const void*>;
+  func_variant_t func_variant;
   dim3 gridDim;
   dim3 blockDim;
-  size_t sharedMem;
+  size_t sharedMem = 0;
 
   // Vector of pointers to the arg_tuple which saves arguments in a typed-erased way
-  ::std::vector<void*> args_ptr;
+  // Mutable so that launch can be const
+  mutable ::std::vector<void*> args_ptr;
```

Comment:
Suggestion: This does not make sense to me. Kernel arguments should not be modified by a kernel launch in my understanding. I would rather cast `const` away in the launch function, which you already do in other places.

Reply:
@caugonnet worth considering, thanks @bernhardmgruber